python/dgl/nn/pytorch/explain/gnnexplainer.py

```diff
@@ -425,3 +398,395 @@ def explain_graph(self, graph, feat, **kwargs):
+            Edge mask of shape :math:`(E)`, where :math:`E`
+            is the number of edges.
+        Returns
+        -------
+        Tensor
+            Loss value with regularization terms added.
+        """
+        # epsilon for numerical stability
+        eps = 1e-15
+
+        for edge_mask in edge_masks.values():
+            edge_mask = edge_mask.sigmoid()
+            # Edge mask sparsity regularization
+            loss = loss + self.alpha1 * torch.sum(edge_mask)
+            # Edge mask entropy regularization
+            ent = - edge_mask * torch.log(edge_mask + eps) - \
+                (1 - edge_mask) * torch.log(1 - edge_mask + eps)
+            loss = loss + self.alpha2 * ent.mean()
+
+        for feat_mask in feat_masks.values():
+            feat_mask = feat_mask.sigmoid()
+            # Feature mask sparsity regularization
+            loss = loss + self.beta1 * torch.mean(feat_mask)
+            # Feature mask entropy regularization
+            ent = -feat_mask * torch.log(feat_mask + eps) - \
+                (1 - feat_mask) * torch.log(1 - feat_mask + eps)
+            loss = loss + self.beta2 * ent.mean()
+
+        return loss
+
+    def explain_node(self, ntype, node_id, graph, feat, **kwargs):
+        r"""Learn and return a node feature mask and subgraph that play a
+        crucial role to explain the prediction made by the GNN for node
+        :attr:`node_id`.
+        Parameters
+        ----------
+        ntype: str
+            The node type of the node to explain.
+        node_id : int
+            The node to explain.
+        graph : DGLHeteroGraph
+            A heterogeneous graph.
+        feat : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        kwargs : dict
+            Additional arguments passed to the GNN model. Tensors whose
+            first dimension is the number of nodes or edges will be
+            assumed to be node/edge features.
+        Returns
+        -------
+        new_node_id : Tensor
+            The new ID of the input center node.
+        sg : DGLHeteroGraph
+            The subgraph induced on the k-hop in-neighborhood of :attr:`node_id`.
+        feat_mask : Dictionary of Tensor
+            Dictionary of { ntypes: features }
+            Learned feature importance mask of shape :math:`(D)`, where :math:`D` is the
+            feature size. The values are within range :math:`(0, 1)`.
+            The higher, the more important.
+        edge_mask : Dictionary of Tensor
+            Dictionary of { canonical_etypes: features }
+            Learned importance mask of the edges in the subgraph, which is a tensor
+            of shape :math:`(E)`, where :math:`E` is the number of edges in the
+            subgraph. The values are within range :math:`(0, 1)`.
+            The higher, the more important.
```

Comment:
Ditto

Reply:
made changes in the recent commit, can you please take a look and let me know if further improvement is needed in the explanation?